Development

MariuszSolvd opened the pull request:
Finished all(5 test cases) for Desktop and Android Mobile

AnnaSolvd commented:
I would add spaces between WebElement declarations for better readability

MateuszKalenik01 commented:
Seems fine for me

MariuszSolvd commented:
> I would add spaces between WebElement declarations for better readability

Done

MKacprzakSolvd commented:
I have problem with test 5 stablity. It fails on the last assertion - when I checked in debugger it showed that jobList is empty, but when I called `jobListPage.getJobList(jobElements)` it returned non-empty list. Seems to be something related to waiting - when I added Thread.sleep(5000) before [this line](https://github.com/MariuszSolvd/web-automation-java-carina/blob/382a81126e7facac612dd97d8d08d47ce6059fba/src/test/java/com/solvd/WebTest.java#L123) i got 9/10 to 10/10 passes, from 1-2/10. Not sure if this happens on your end, or is there something wrong with my infrastructure. My tests seem to be working relatively well.

MKacprzakSolvd commented:
Beside that, I thing you could improve are those regexes:
https://github.com/MariuszSolvd/web-automation-java-carina/blob/382a81126e7facac612dd97d8d08d47ce6059fba/src/main/java/com/solvd/mapper/JobMapper.java#L28-L29
If you want it to work with \n and \n\r then it would be best to replace them with `.split("\\R")` ([more](https://stackoverflow.com/questions/3445326/regex-in-java-how-to-deal-with-newline))

MKacprzakSolvd commented:
Beside thoses I don't see any issues

MariuszSolvd commented:
> Beside that, I thing you could improve are those regexes:
> 
> https://github.com/MariuszSolvd/web-automation-java-carina/blob/382a81126e7facac612dd97d8d08d47ce6059fba/src/main/java/com/solvd/mapper/JobMapper.java#L28-L29
> 
> 
> If you want it to work with \n and \n\r then it would be best to replace them with `.split("\\R")` ([more](https://stackoverflow.com/questions/3445326/regex-in-java-how-to-deal-with-newline))

Thanks a lot! I will improve it, first thing at morning.



> I have problem with test 5 stablity. It fails on the last assertion - when I checked in debugger it showed that jobList is empty, but when I called `jobListPage.getJobList(jobElements)` it returned non-empty list. Seems to be something related to waiting - when I added Thread.sleep(5000) before [this line](https://github.com/MariuszSolvd/web-automation-java-carina/blob/382a81126e7facac612dd97d8d08d47ce6059fba/src/test/java/com/solvd/WebTest.java#L123) i got 9/10 to 10/10 passes, from 1-2/10. Not sure if this happens on your end, or is there something wrong with my infrastructure. My tests seem to be working relatively well.

It's weird. It worked on my end 9/10. I didn't just when I had my emulator in backgroud due to slow loading site. If you will go to [HERE](https://github.com/MariuszSolvd/web-automation-java-carina/blob/c481f998db6bad3fd2c45cf23ccea0489a764dba/src/main/java/com/solvd/pages/common/admin/AbstractAdminPage.java#L14) you'll see I have set loader so it(and in my case does) wait for load element at site which I set in config to BY_VISIBILITY. Moreover [here](https://github.com/MariuszSolvd/web-automation-java-carina/blob/382a81126e7facac612dd97d8d08d47ce6059fba/src/test/java/com/solvd/WebTest.java#L121) I assert if it's open again. Anyways, again thanks a lot for not only ready my code, but testing it by yourself! I'll look closely at TC 5 tomorrow as well :)

MKacprzakSolvd commented:
Hope it helps. Maybe it's something on my part. One thing I forgot to mention about waiting: I measured execution time of [that asssertPageOpened](https://github.com/MariuszSolvd/web-automation-java-carina/blob/382a81126e7facac612dd97d8d08d47ce6059fba/src/test/java/com/solvd/WebTest.java#L121) and it seems to be waiting every time (around 4 to 5 seconds in my case), even in failed tests.

I took quick look at the website, and it seems that there is a state when `//div[@role ='table']` element is present, but not filled in. Maybe this causes an issue?

<img width="1306" alt="Screenshot 2024-06-26 at 09 35 09" src="https://github.com/MariuszSolvd/web-automation-java-carina/assets/170347259/e03d701b-3acc-45c6-9b9c-94863b4571f0">

MariuszSolvd commented:
> Hope it helps. Maybe it's something on my part. One thing I forgot to mention about waiting: I measured execution time of [that asssertPageOpened](https://github.com/MariuszSolvd/web-automation-java-carina/blob/382a81126e7facac612dd97d8d08d47ce6059fba/src/test/java/com/solvd/WebTest.java#L121) and it seems to be waiting every time (around 4 to 5 seconds in my case), even in failed tests.
> 
> I took quick look at the website, and it seems that there is a state when `//div[@role ='table']` element is present, but not filled in. Maybe this causes an issue?
> 
> <img alt="Screenshot 2024-06-26 at 09 35 09" width="1306" src="https://private-user-images.githubusercontent.com/170347259/343053809-e03d701b-3acc-45c6-9b9c-94863b4571f0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzODc4NDYsIm5iZiI6MTcxOTM4NzU0NiwicGF0aCI6Ii8xNzAzNDcyNTkvMzQzMDUzODA5LWUwM2Q3MDFiLTNhY2MtNDVjNi05YjljLTk0ODYzYjQ1NzFmMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYyNlQwNzM5MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lY2M0Njk1OTQ5ZjcxYTYyMDU1NWVlODkyYmRmNjEyYjBiMTU4MDk0MDllNmQyNTczMzhiMGExYjcxMjk4Njk2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.9JE8SbHG27d5-tEfuMO3iwEUNWS72-EWyzNXtBpceQE">

Awesome! I will change assertion to different element. Seems like you might be right 'cuz it's what I waiting for. Let me change that I will test it and when i push, I will ask you to check again, if it's not a problem for you. So you can approve it on your side.

MKacprzakSolvd commented:
I checked it, and it  still fails in around 90% when executing test 5. If the fixed waiting for element is for this issue then I'm failing to see it. Beside that exerything seems good

MKacprzakSolvd commented on src/main/java/com/solvd/pages/common/admin/AddJobPageBase.java:
```diff
@@ -1,18 +1,45 @@
+    }
 
-    public abstract void clickSaveButton();
+    public void inputDescription(String description) {
```
Wouldn't be better to make this private / protected, since it's not used anywhere outside this method? It seems to me that it would make using this class in test easier (less public methods to choose from). Same for inputJobTitle

MariuszSolvd replied:
Yes, fixed, thanks!

MariuszSolvd replied:
OK I get rid of this method totally. Since I use it just once, create all method seems too much. I just use field direcrly now.

MKacprzakSolvd commented on src/main/java/com/solvd/pages/common/admin/AdminPageBase.java:
```diff
@@ -1,16 +1,31 @@
+    }
 
-    public abstract void clickJobTitleButton();
+    public void clickJobTitleButton() {
```
Wouldn't be better to make this private / protected, since it's not used anywhere outside this method? It seems to me that it would make using this class in test easier (less public methods to choose from). Same for clickJobButton

MariuszSolvd replied:
Again, yes, you're right. I deleted both methods, and use now just a field with call method click.

MKacprzakSolvd commented on src/main/java/com/solvd/pages/common/admin/AdminPageBase.java:
```diff
@@ -1,16 +1,31 @@
+    }
 
-    public abstract JobListPageBase getJobListPage();
+    public JobListPageBase getJobListPage() {
```
Wouldn't goToJobListPage be a better name, since this method navigates to different page, not returns some field of itself? Sorry if I'm too nitpicky

MariuszSolvd replied:
We had call about it, I explained to you why I think this way is better :)

MKacprzakSolvd commented on src/test/java/com/solvd/WebTest.java:
```diff
@@ -109,9 +105,9 @@ public void shouldAddJobTitle() {
         LoginService loginService = new LoginService();
```
Test 5 is not stable

MariuszSolvd replied:
Now should be stable.

MKacprzakSolvd replied:
Can confirm, works stable on all platforms

MKacprzakSolvd reviewed (commented):
Only thing that I think is needed for passing review is making test 5 more stable, rest are just suggestions, though I would be interested in hearing your feedback on them.

Two ideas to improve your code:
- On login page it might be better to have two methods: `public LoginPageBase loginIncorrectly(...)` and `public SomeOtherPageBase loginCorrectly(...)` that way you avoid having to call initPage after login.
- Change `page.clickMenuButtonByHref(SOME_PAGE_HREF)` to `page.goToPageFromMenu(MenuOptions.SOME_PAGE)` and have MenuOptions enum have href field. I think that would improve readability.

MariuszSolvd commented:
> Only thing that I think is needed for passing review is making test 5 more stable, rest are just suggestions, though I would be interested in hearing your feedback on them.
> 
> Two ideas to improve your code:
> 
> * On login page it might be better to have two methods: `public LoginPageBase loginIncorrectly(...)` and `public SomeOtherPageBase loginCorrectly(...)` that way you avoid having to call initPage after login.
> * Change `page.clickMenuButtonByHref(SOME_PAGE_HREF)` to `page.goToPageFromMenu(MenuOptions.SOME_PAGE)` and have MenuOptions enum have href field. I think that would improve readability.

First point, if I understand correctly, you want me to do some kind of service which I already have and use it as a correct login and incorrect. It would be bad idea since when I do tests, it's better to include steps in test itself instead of create service(Sergye told us this on laba). When this functionality is as Preconditions then you can delegate it to service and your solution would be good just to not make it inside of PageObject class just service. It's what Sergey told us on laba.

Second point, change name isn't a big deal. I pick it 'cuz it's action to click on button in menu and it takes href as an argument, so it describes everything. About ENUM it's good option, was thinking about it to be honest but Andrei show me solution that they use on his project which is Interface with URLs and passing them as an argument directly. That's why I implemented this way, but, yeah ENUMS seems right way to do it, and was on my mid.

MKacprzakSolvd commented on src/main/java/com/solvd/pages/common/components/LeftMenu.java:
```diff
@@ -20,9 +20,8 @@ public LeftMenu(WebDriver webDriver, SearchContext searchContext) {
         return menuButtons.stream()
-                .filter(button -> button.getAttribute("href").equals(href))
+                .filter(button -> button.getAttribute("href").contains(href))
                 .findFirst()
```
Wouldn't be better to use `findAny()` here, since you just want to find one button?

MariuszSolvd replied:
Done

MKacprzakSolvd commented on src/main/java/com/solvd/pages/desktop/PageWithLeftMenuDesktop.java:
```diff
@@ -0,0 +1,27 @@
+import org.openqa.selenium.WebDriver;
+
+
+@Getter
```
Why do you need `@Getter` here?

MariuszSolvd replied:
I don't need it. I left it probably during refactoring. Fixed

whetman commented on src/main/java/com/solvd/pages/mobile/admin/AddJobPageMobile.java:
```diff
@@ -0,0 +1,25 @@
+import org.openqa.selenium.support.ui.ExpectedConditions;
+
+
+public  class AddJobPageMobile extends AddJobPageBase {
```
did you want to make this class abstract? just asking, since you didn't add device type and the space between public and class is big and it looks similar to your other abstract clasess

whetman reviewed (approved):
Looks good to me :)

MKacprzakSolvd commented on src/main/java/com/solvd/service/LoginService.java:
```diff
@@ -1,19 +1,18 @@
 
 
-    public DashboardPageBase successfulLogin() {
+    public PageWithLeftMenuBase successfulLogin() {
```
I think it would be a good idea to add method `public <T> T successfulLogin(PageType or url or sth)` that will will do the redirecting, that way you will avoid repetition, since after every usage of this method you do the same steps:

https://github.com/MariuszSolvd/web-automation-java-carina/blob/5583453dece8a731b4e1996eca5d36acfaf46e06/src/test/java/com/solvd/WebTest.java#L62-L67

https://github.com/MariuszSolvd/web-automation-java-carina/blob/5583453dece8a731b4e1996eca5d36acfaf46e06/src/test/java/com/solvd/WebTest.java#L85-L90

https://github.com/MariuszSolvd/web-automation-java-carina/blob/5583453dece8a731b4e1996eca5d36acfaf46e06/src/test/java/com/solvd/WebTest.java#L111-L116

MariuszSolvd replied:
I could do class like this, I know it. Something like:

public <T extends PageWithLeftMenuBase> T successfulLogin(Class<T> page) {
        LoginPageBase loginPage = initPage(getDriver(), LoginPageBase.class);
        loginPage.open();
        loginPage.logIn(R.TESTDATA.get("correct_user"), R.TESTDATA.get("correct_password"));

        return initPage(getDriver(), page);
    }
    
 but what I would pass in page as an argument? PageWithLeftMenuBase which will be repetition. Keep in mind that I can't predict which page will be opened after each test, so if I will have to pass url or page(I think page would be most suitable) still I would have to know which one it would be, which I do not know. So current solution seems for me reasonable. Since I know that each page has basic navigation, I just use it, and then redirect to site I will know it's this site and will have this implementation.

MariuszSolvd replied:
I know what you mean. Like this way, right?

```java
public <T extends PageWithLeftMenuBase> T successfulLogin(Class<T> page) {
        LoginPageBase loginPage = initPage(getDriver(), LoginPageBase.class);
        loginPage.open();
        loginPage.logIn(R.TESTDATA.get("correct_user"), R.TESTDATA.get("correct_password"));

        return initPage(getDriver(), page);
}
```
It can't be done 'cuz, I don't know which site I will get on. So which page I should pass there, or url? If I will pass AddEmployeePageBase.class, I ain't sure I will be redirected after login to this site. So this solution is better, more optimal. Since PageWithLeftMenuBase it's a CORE, BASE of all pages after login, and it has basic functions implemented. Then from this site I can redirect to others and then I am sure I am on this page.

MKacprzakSolvd replied:
More like:
``` java
public <T extends PageWithLeftMenuBase> T successfulLogin(Class<T> page, String href) {
        LoginPageBase loginPage = initPage(getDriver(), LoginPageBase.class);
        loginPage.open();
        loginPage.logIn(R.TESTDATA.get("correct_user"), R.TESTDATA.get("correct_password"));
        PageWithLeftMenuBase pageWithLeftMenuBase = initPage(...);

        pageWithLeftMenuBase.clickMenuButtonByHref(ADMIN_BUTTON_HREF); 
        T selectedPage = initPage(getDriver(), page); 
        selectedPage.assertPageOpened();

        return selectedPage;
}
```
or use some kind of enum that have href and page class, or method to deduce page href by class, or just call openPage:
``` java
    ...
    loginPage.logIn(R.TESTDATA.get("correct_user"), R.TESTDATA.get("correct_password"));
    T selectedPage = initPage(getDriver(), page); 
    selectedPage.open();
    ....
```

MariuszSolvd replied:
Improved as we discussed on our call, I just had to add assertion after login, works good! :)

```java
public <T extends PageWithLeftMenuBase> T successfulLogin(Class<T> page) {
        LoginPageBase loginPage = initPage(getDriver(), LoginPageBase.class);
        loginPage.open();
        loginPage.logIn(R.TESTDATA.get("correct_user"), R.TESTDATA.get("correct_password"));
        PageWithLeftMenuBase pageWithLeftMenuBase = initPage(getDriver(), PageWithLeftMenuBase.class);
        pageWithLeftMenuBase.assertPageOpened();

        T targetPage = initPage(getDriver(), page);
        targetPage.open();
        targetPage.assertPageOpened();
        return  targetPage;
    }
    ```

MKacprzakSolvd commented on src/main/resources/_config-desktop.properties:
```diff
@@ -0,0 +1,2 @@
+#==============DESKTOP======================
+capabilities.platformName=mac
```
Making tests work on desktop on my infrastructure required adding:
`capabilities.browserName=chrome` here, otherwise it sometimes tried to connect to safari browser and failed, but that is probably partially caused by the way I have my infrastructure set up. Nevertheless it might be a good idea to add this line to explicitly state what browser you want to use.

MariuszSolvd replied:
That's correct. After refactoring code, I move capabilities.browserName to each config. Fixed, thanks!

MKacprzakSolvd commented on src/test/java/com/solvd/WebTest.java:
```diff
@@ -73,53 +72,59 @@ public void shouldAddEmployee() {
+
     }
 
     //Test case 4
```
T4 for iPhone:
Sometimes it does works, sometimes it doesn't work on my infrastructure, usually ends with one of the following exceptions:
- `org.openqa.selenium.NoSuchElementException: no such element By.xpath: //div[@class = 'oxd-table-filter-header']//button`
  ```
  at com.solvd.pages.mobile.pim.PimPageMobile.clickOpenSearchButton(PimPageMobile.java:64)
  at com.solvd.pages.mobile.pim.PimPageMobile.inputIdEmployee(PimPageMobile.java:45)
  at com.solvd.WebTest.shouldDeleteEmployee(WebTest.java:99)
  ```
- `org.openqa.selenium.NoSuchElementException: no such element By.xpath: //button[contains(@class, 'label-danger')]`
  ```
  at com.solvd.pages.common.pim.PimPageBase.clickDeleteConfirmationButton(PimPageBase.java:53)
  at com.solvd.WebTest.shouldDeleteEmployee(WebTest.java:97)
  ```
- `org.openqa.selenium.InvalidElementStateException: Element is not currently interactable and may not be manipulated`
  ```
  at com.solvd.pages.mobile.pim.PimPageMobile.inputIdEmployee(PimPageMobile.java:49)
  at com.solvd.WebTest.shouldDeleteEmployee(WebTest.java:99)
  ```
- `org.openqa.selenium.WebDriverException: Element is not currently visible and may not be manipulated`
  ```
  at com.solvd.pages.mobile.pim.PimPageMobile.clickSearchEmployeeButton(PimPageMobile.java:40)
  at com.solvd.WebTest.shouldDeleteEmployee(WebTest.java:100)
  ```

MariuszSolvd replied:
It always for for me. As we discussed before, it goes super fast on my end and, I thought it doesn't work as well but it did. I will try to improve it, add some waiters and then we will see.

MariuszSolvd replied:
Ok I added two waiters, none in my opinion is needed. I wonder right now if I should delete them, cuz it works similar on my end with them or without them, just test is little bit longer. I might leave them tho to increase stability and be sure. Again in my end, I do not see things happens, but in reality, it does. I can't see clicks(delete, confirm delete, search, sometimes even typing id) but it happens somehow 'cuz I noticed user deleted with credentials that shown in my log. One thing that I could improve is redirection after login. This one seems to be some kinda issue on my end.

MKacprzakSolvd commented on src/test/java/com/solvd/WebTest.java:
```diff
@@ -73,53 +72,59 @@ public void shouldAddEmployee() {
+
     }
 
     //Test case 4
```
T4 for iPhone:
Fails every time, there is a red popup on page with text: "Error\n cannot be deleted"
Exception: `org.openqa.selenium.NoSuchElementException: no such element By.xpath: //button[contains(@class, 'label-danger')]`
Trace:
```
...
at com.solvd.pages.common.pim.PimPageBase.clickDeleteConfirmationButton(PimPageBase.java:53)
at com.solvd.WebTest.shouldDeleteEmployee(WebTest.java:97)
...
```

MariuszSolvd replied:
Ignore this one please. It will happen when there is no users added by customer. There is one hardcoded user, that cannot be deleted and always will be there. So if there is no users added, my test will try to delete this one(as it just pick first user from list), and it cannot be deleted. We discussed this case, on call, already. I could implement adding user and delete one I added, but it does not happen so often, so I decided to move on, since my priority right now is to start native app testing.

MKacprzakSolvd commented on src/main/java/com/solvd/pages/mobile/ios/admin/AddJobPageIOs.java:
```diff
@@ -0,0 +1,15 @@
+import org.openqa.selenium.WebDriver;
+
+@DeviceType(pageType = DeviceType.Type.IOS_PHONE, parentClass = AddJobPageBase.class)
+public class AddJobPageIOs extends AddJobPageAndroid {
```
Shouldn't this class inherit from `AddJobPageMobile`?
```suggestion
public class AddJobPageIOs extends AddJobPageMobile {
```

MariuszSolvd replied:
Yes, thanks. Fixed.

MKacprzakSolvd commented on src/main/java/com/solvd/pages/mobile/pim/AddEmployeePageMobile.java:
```diff
@@ -1,40 +1,47 @@
-        pressBack();
-        swipe(saveButton);
-        saveButton.click();
+    public EmployeePageBase addEmployee(Employee employee) {
```
I don't understand the point of overriding this method, since it is implemented exactly like in the parent class (and also it's the same in AddEmployeePageDesktop):
This implementation (AddEmployeePageMobile):
https://github.com/MariuszSolvd/web-automation-java-carina/blob/5583453dece8a731b4e1996eca5d36acfaf46e06/src/main/java/com/solvd/pages/mobile/pim/AddEmployeePageMobile.java#L24-L32
Parent implementation (AddEmployeePageBase):
https://github.com/MariuszSolvd/web-automation-java-carina/blob/5583453dece8a731b4e1996eca5d36acfaf46e06/src/main/java/com/solvd/pages/common/pim/AddEmployeePageBase.java#L49-L56
AddEmployeePageDesktop:
https://github.com/MariuszSolvd/web-automation-java-carina/blob/5583453dece8a731b4e1996eca5d36acfaf46e06/src/main/java/com/solvd/pages/desktop/pim/AddEmployeePageDesktop.java#L36-L44

MariuszSolvd replied:
Correct, I left it just in AddEmployeePageBase.java

MKacprzakSolvd commented on src/main/java/com/solvd/pages/mobile/pim/EmployeePageMobile.java:
```diff
@@ -38,8 +39,14 @@ public void clickMenuButtonByHref(String href) {
+    private String getValueWrapper(ExtendedWebElement element) {
         JavascriptExecutor js = (JavascriptExecutor) this.getDriver();
-        return (String) js.executeScript("return arguments[0].value;", element);
+        return (String) js.executeScript("return arguments[0].value;", element.getElement());
```
Wouldn't be simpler to just use `element.getAttribute("value")` instead of `JavascriptExecutor`?

MariuszSolvd replied:
Nah, tried it. Somehow it does not work for mobile. It works fine in Desktop and there you can see it, but, it doesn't work for mobile.

MKacprzakSolvd commented on src/main/java/com/solvd/pages/mobile/pim/PimPageMobile.java:
```diff
@@ -3,27 +3,35 @@
+        super(webDriver);
+    }
 
     @FindBy(xpath = "//div[@class = 'oxd-table-filter-header']//button")
```
I think it would be better to use contains, when checking if element have some class. That way you locator will be more stable, and will work even if some other classes will be added to the element

MariuszSolvd replied:
Done

MKacprzakSolvd commented on src/main/java/com/solvd/pages/mobile/pim/PimPageMobile.java:
About `inputIdEmployee()` method, I thnk that it might be usefull to extract this:
https://github.com/MariuszSolvd/web-automation-java-carina/blob/5583453dece8a731b4e1996eca5d36acfaf46e06/src/main/java/com/solvd/pages/mobile/pim/PimPageMobile.java#L47-L48
into some utility class like inputClearer, that way you could use it here:
https://github.com/MariuszSolvd/web-automation-java-carina/blob/5583453dece8a731b4e1996eca5d36acfaf46e06/src/main/java/com/solvd/pages/mobile/ios/pim/AddEmployeePageIOs.java#L24-L28
and here:
https://github.com/MariuszSolvd/web-automation-java-carina/blob/5583453dece8a731b4e1996eca5d36acfaf46e06/src/main/java/com/solvd/pages/mobile/android/pim/AddEmployeePageAndroid.java#L25-L29
Not sure that it will work, just an idea

MariuszSolvd replied:
If I recall correctly I tried at first, and it didn't work. I remember that I was trying to do it on mobile(I think iOS) by JavascriptExecutor and it didn't work so I pick this...kinda primitive way to do it.

MKacprzakSolvd commented on src/main/java/com/solvd/pages/common/pim/PimPageBase.java:
About `inputIdEmployee()` method, I think that you could move it's implementation to this class and just add override to PimPageMobile that calls `clickOpenSearchButton();` before super method
PimPageBase:
https://github.com/MariuszSolvd/web-automation-java-carina/blob/5583453dece8a731b4e1996eca5d36acfaf46e06/src/main/java/com/solvd/pages/common/pim/PimPageBase.java#L50
PimPageMobile:
https://github.com/MariuszSolvd/web-automation-java-carina/blob/5583453dece8a731b4e1996eca5d36acfaf46e06/src/main/java/com/solvd/pages/mobile/pim/PimPageMobile.java#L43-L50
PimPageDesktop:
https://github.com/MariuszSolvd/web-automation-java-carina/blob/5583453dece8a731b4e1996eca5d36acfaf46e06/src/main/java/com/solvd/pages/desktop/pim/PimPageDesktop.java#L38-L44

MariuszSolvd replied:
Was thinking about it, but I pick this way 'cuz, it's more readable for me. Like I have an abstract method and then implementation for each Desktop and Mobile. Doesn't change much but seems for me more readable since implementation are under classes with names Desktop and Mobile.

MariuszSolvd replied:
As we talk during our call, you convinced me that your suggestion is better, so I changed it.

MKacprzakSolvd reviewed (commented):
Other issues:
1. Error on android in LoadingPageAndroid (Session not known)
   Replacing `pressBack();` `hideKeyboard();` in LoginPageAndroid seems to fix the issue for me, though it might be the problem with my infrastructure.
   Error:
   ```
   org.openqa.selenium.NoSuchSessionException: The session identified by 1563c78a-5ef7-4766-947b-6629c83a097d is not known
   Build info: version: '4.13.0', revision: 'ba948ece5b*'
   System info: os.name: 'Mac OS X', os.arch: 'aarch64', os.version: '14.5', java.version: '21.0.3'
   Driver info: io.appium.java_client.android.AndroidDriver
   Command: [1158f65e-3f76-43c7-b478-b5e4633d29ab, executeScript {script=mobile: pressKey, args=[{keycode=4}]}]
   Capabilities {appium:automationName: UiAutomator2, appium:databaseEnabled: false, appium:desired: {appActivity: com.android.chrome/com.goog..., appPackage: com.android.chrome, automationName: UiAutomator2, browserName: chrome, deviceName: emulator-5554, deviceType: phone, language: en, locale: US, platformName: ANDROID}, appium:deviceApiLevel: 34, appium:deviceManufacturer: Google, appium:deviceModel: sdk_gphone64_arm64, appium:deviceName: emulator-5554, appium:deviceScreenDensity: 440, appium:deviceScreenSize: 1080x2220, appium:deviceType: phone, appium:deviceUDID: emulator-5554, appium:javascriptEnabled: true, appium:language: en, appium:locale: US, appium:locationContextEnabled: false, appium:networkConnectionEnabled: true, appium:pixelRatio: 2.75, appium:platformVersion: 14, appium:statBarHeight: 66, appium:takesScreenshot: true, appium:viewportRect: {height: 2154, left: 0, top: 66, width: 1080}, appium:warnings: {}, appium:webStorageEnabled: false, browserName: chrome, platformName: ANDROID, se:bidiEnabled: false, se:cdp: ws://192.168.0.15:4444/sess...}
   Session ID: 1158f65e-3f76-43c7-b478-b5e4633d29ab
   ...
   at com.zebrunner.carina.utils.android.IAndroidUtils.pressBack(IAndroidUtils.java:194)
   at com.solvd.pages.mobile.android.LoginPageAndroid.logIn(LoginPageAndroid.java:20)
   at com.solvd.WebTest.shouldLogin(WebTest.java:43)
   ...
   ```
   I ran it multiple times, each time the same error.
   (Only tested Test 1 and Test 2, not sure about other tests, but I suspect it would be the same)
2.  [Here](https://github.com/MariuszSolvd/web-automation-java-carina/blob/5583453dece8a731b4e1996eca5d36acfaf46e06/src/main/java/com/solvd/pages/common/pim/AddEmployeePageBase.java#L25) I think it would be better to use `contains`, when checking if element have some class. That way you locator will be more stable, and will work even if some other classes will be added to the element
3.

MKacprzakSolvd commented on src/main/java/com/solvd/app/ios/BrowserPageIOs.java:
```diff
@@ -0,0 +1,32 @@
+    @Override
+    public void openNewTab() {
+        tabButton.click();
+        addNewTabButton.click();
```
`switchTab` test for iOS fails here
exception:
```
org.openqa.selenium.NoSuchElementException: no such element By.xpath: //XCUIElementTypeButton[@name="NewTabButton"]
```
calls:
```
...
at com.solvd.app.ios.BrowserPageIOs.openNewTab(BrowserPageIOs.java:25)
at com.solvd.ContextTest.switchTab(ContextTest.java:27)
...
```

MariuszSolvd commented:
> Other issues:
> 
> 1. Error on android in LoadingPageAndroid (Session not known)
>    Replacing `pressBack();` `hideKeyboard();` in LoginPageAndroid seems to fix the issue for me, though it might be the problem with my infrastructure.
>    Error:
>    ```
>    org.openqa.selenium.NoSuchSessionException: The session identified by 1563c78a-5ef7-4766-947b-6629c83a097d is not known
>    Build info: version: '4.13.0', revision: 'ba948ece5b*'
>    System info: os.name: 'Mac OS X', os.arch: 'aarch64', os.version: '14.5', java.version: '21.0.3'
>    Driver info: io.appium.java_client.android.AndroidDriver
>    Command: [1158f65e-3f76-43c7-b478-b5e4633d29ab, executeScript {script=mobile: pressKey, args=[{keycode=4}]}]
>    Capabilities {appium:automationName: UiAutomator2, appium:databaseEnabled: false, appium:desired: {appActivity: com.android.chrome/com.goog..., appPackage: com.android.chrome, automationName: UiAutomator2, browserName: chrome, deviceName: emulator-5554, deviceType: phone, language: en, locale: US, platformName: ANDROID}, appium:deviceApiLevel: 34, appium:deviceManufacturer: Google, appium:deviceModel: sdk_gphone64_arm64, appium:deviceName: emulator-5554, appium:deviceScreenDensity: 440, appium:deviceScreenSize: 1080x2220, appium:deviceType: phone, appium:deviceUDID: emulator-5554, appium:javascriptEnabled: true, appium:language: en, appium:locale: US, appium:locationContextEnabled: false, appium:networkConnectionEnabled: true, appium:pixelRatio: 2.75, appium:platformVersion: 14, appium:statBarHeight: 66, appium:takesScreenshot: true, appium:viewportRect: {height: 2154, left: 0, top: 66, width: 1080}, appium:warnings: {}, appium:webStorageEnabled: false, browserName: chrome, platformName: ANDROID, se:bidiEnabled: false, se:cdp: ws://192.168.0.15:4444/sess...}
>    Session ID: 1158f65e-3f76-43c7-b478-b5e4633d29ab
>    ...
>    at com.zebrunner.carina.utils.android.IAndroidUtils.pressBack(IAndroidUtils.java:194)
>    at com.solvd.pages.mobile.android.LoginPageAndroid.logIn(LoginPageAndroid.java:20)
>    at com.solvd.WebTest.shouldLogin(WebTest.java:43)
>    ...
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    I ran it multiple times, each time the same error.
>    (Only tested Test 1 and Test 2, not sure about other tests, but I suspect it would be the same)
> 2. [Here](https://github.com/MariuszSolvd/web-automation-java-carina/blob/5583453dece8a731b4e1996eca5d36acfaf46e06/src/main/java/com/solvd/pages/common/pim/AddEmployeePageBase.java#L25) I think it would be better to use `contains`, when checking if element have some class. That way you locator will be more stable, and will work even if some other classes will be added to the element

1. I don't have this issue on my end. I just check it and it works for Android as well.
2. Good catch

MKacprzakSolvd reviewed (approved):
LGTM

whetman reviewed (approved):
Everything works fine

whetman reviewed (approved):
Looks good

MKacprzakSolvd reviewed (approved):
LGTM

whetman reviewed (approved):
LGTM

whetman reviewed (approved):
👍
